File: src/test/java/ev/projects/WallTest.java
package ev.projects;

import org.junit.jupiter.api.Test;

import java.util.List;

import static ev.projects.PatternLineTest.patternLines;
import static org.junit.jupiter.api.Assertions.*;

public class WallTest {

    @Test
    void tileGetsPlacedOnTheWallIfPatternLineIsFilled() {
        Wall wall = new Wall();
        PatternLine[] patternLines = patternLines();
        patternLines[4].add(Tile.RED, 5);
        Game game = new Game(List.of(new Player(new Board(patternLines, wall, new Floor()))));

        game.executeWallTilingPhase();

        assertTrue(wall.alreadyHas(Tile.RED));
        assertEquals(0, patternLines[4].tileCount());
    }

    @Test
    void tileDoesNotGetPlacedOnTheWallIfPatternLineIsNotFilled() {
        Wall wall = new Wall();
        PatternLine[] patternLines = patternLines();
        patternLines[4].add(Tile.RED, 4);
        Game game = new Game(List.of(new Player(new Board(patternLines, wall, new Floor()))));

        game.executeWallTilingPhase();

        assertFalse(wall.alreadyHas(Tile.RED));
        assertEquals(4, patternLines[4].tileCount());
    }

    @Test
    void cantAddColourToPatternLineIfWallHasThatColour() {
        Wall wall = new Wall();
        Player player = new PlayerMother().newPlayer(wall);
        player.takeTiles(List.of(Tile.RED));
        wall.add(Tile.RED, 4);

        assertThrows(ActionNotAllowedException.class, () -> player.addTileToPatternLine(1, 4));
    }

    @Test
    void playerScoresAPointWhenPlacingATileOnEmptyWall() {
        Wall wall = new Wall();
        PatternLine[] patternLines = patternLines();
        patternLines[0].add(Tile.RED, 1);
        Player player = new Player(new Board(patternLines, wall, new Floor()));

        player.moveTileToWall();

        assertEquals(1, player.score());
    }

    @Test
    void playerScoresPointsWhenPlacingATileOnWall() {
        Wall wall = new Wall();
        PatternLine[] patternLines = patternLines();
        patternLines[2].add(Tile.YELLOW, 3);
        Player player = new Player(new Board(patternLines, wall, new Floor()));
        wall.add(Tile.BLACK, 2);
        wall.add(Tile.WHITE, 2);
        wall.add(Tile.BLUE, 2);
        wall.add(Tile.BLACK, 0);
        wall.add(Tile.RED, 1);
        int currentScore = player.score();

        player.moveTileToWall(2);

        assertEquals(7, player.score() - currentScore);
    }

    @Test
    void playerScoresAPointWhenPlacingATileOnWall() {
        Wall wall = new Wall();
        PatternLine[] patternLines = patternLines();
        patternLines[2].add(Tile.YELLOW, 3);
        Player player = new Player(new Board(patternLines, wall, new Floor()));
        wall.add(Tile.BLACK, 2);
        wall.add(Tile.WHITE, 2);
        wall.add(Tile.BLACK, 0);
        int currentScore = player.score();

        player.moveTileToWall(2);

        assertEquals(1, player.score() - currentScore);
    }
}

File: python/test/wall_test.py
import unittest

from flask import Flask

from src.lid import Lid
from src.wall import Wall
from src.pattern_line import PatternLine
from src.board import Board
from src.floor import Floor
from src.player import Player
from src.game import Game
from src.tile import Tile
from src.center import Center
from src.factory_display import FactoryDisplay
from src.action_not_allowed_exception import ActionNotAllowedException
from player_mother import PlayerMother


class TestWall(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.app = Flask(__name__)
        cls.app_context = cls.app.app_context()
        cls.app_context.push()

    @classmethod
    def tearDownClass(cls):
        cls.app_context.pop()

    def test_tile_gets_placed_on_wall_if_pattern_line_is_filled(self):
        wall = Wall()
        pattern_lines = self.pattern_lines()
        pattern_lines[4].add(Tile.RED, 5)
        game = Game([Player(Board(pattern_lines, wall, Floor()))])

        game.execute_wall_tiling_phase()

        self.assertTrue(wall.already_has(Tile.RED, 4))
        self.assertEqual(0, pattern_lines[4].tile_count)

    def test_tile_does_not_get_placed_on_wall_if_pattern_line_is_not_filled(self):
        wall = Wall()
        pattern_lines = self.pattern_lines()
        pattern_lines[4].add(Tile.RED, 4)
        game = Game([Player(Board(pattern_lines, wall, Floor()))])

        game.execute_wall_tiling_phase()

        self.assertFalse(wall.already_has(Tile.RED, 4))
        self.assertEqual(4, pattern_lines[4].tile_count)

    def test_cant_add_colour_to_pattern_line_if_wall_has_that_colour(self):
        wall = Wall()
        player = PlayerMother.new_player(wall=wall)
        wall.add(Tile.RED, 4)

        with self.assertRaises(ActionNotAllowedException):
            player.take_tiles_from_factory(
                FactoryDisplay(Center(), [Tile.RED, Tile.RED, Tile.RED, Tile.BLUE]), Tile.RED, 0, 4
            )

    def test_player_scores_a_point_when_placing_tile_on_empty_wall(self):
        wall = Wall()
        pattern_lines = self.pattern_lines()
        pattern_lines[0].add(Tile.RED, 1)
        player = Player(Board(pattern_lines, wall, Floor()))

        player.move_tiles_to_wall(Lid())

        self.assertEqual(1, player.score)

    def test_player_scores_points_when_placing_tile_on_wall(self):
        wall = Wall()
        pattern_lines = self.pattern_lines()
        pattern_lines[2].add(Tile.YELLOW, 3)
        player = Player(Board(pattern_lines, wall, Floor()))
        wall.add(Tile.BLACK, 2)
        wall.add(Tile.WHITE, 2)
        wall.add(Tile.BLUE, 2)
        wall.add(Tile.BLACK, 0)
        wall.add(Tile.RED, 1)
        current_score = player.score

        player.move_tiles_to_wall(Lid())

        self.assertEqual(7, player.score - current_score)

    def test_player_scores_a_point_when_placing_tile_on_wall(self):
        wall = Wall()
        pattern_lines = self.pattern_lines()
        pattern_lines[2].add(Tile.YELLOW, 3)
        player = Player(Board(pattern_lines, wall, Floor()))
        wall.add(Tile.BLACK, 2)
        wall.add(Tile.WHITE, 2)
        wall.add(Tile.BLACK, 0)
        current_score = player.score

        player.move_tiles_to_wall(Lid())

        self.assertEqual(1, player.score - current_score)

    @staticmethod
    def pattern_lines():
        return [PatternLine(1), PatternLine(2), PatternLine(3), PatternLine(4), PatternLine(5)]
